(FACT-1873) Use builtin fact if custom facts don't resolve or have 0 wt.

ekinanp opened the pull request:
Previously, if any builtin facts have conflicting custom facts that all
fail to resolve, and at least one of these custom facts has a weight >
0, Facter will not report a value for that fact. This is undesirable
behavior because we always want to fall back to the builtin value if
all custom facts fail to resolve. Note that the "weight > 0" part is
important because Facter _will_ use the builtin value if all custom
facts have 0 weight.

This commit falls back to the builtin value if all custom facts fail
to resolve, which fixes the reported bug in the ticket.

ekinanp commented on lib/src/ruby/fact.cc:
```diff
@@ -120,6 +120,21 @@ namespace facter { namespace ruby {
                 _value = value;
                 _weight = weight;
+
+                if (! ruby.is_nil(_value)) {
```
Some questions/comments on this:
* Currently when all custom facts have weight 0, we use the builtin fact value (see [here](https://github.com/puppetlabs/facter/blob/master/lib/src/ruby/fact.cc#L87-L99). Is this desired behavior?
* If ^ is the desired behavior, then this code will break that consistency when all of our custom facts with weight > 0 fail to resolve and one custom fact with weight == 0 resolves. Would that be bad?

McdonaldSeanp replied:
According to the docs facts should always have a positive integer value (i.e. not 0): https://docs.puppet.com/facter/3.0/fact_overview.html

So as a user I wouldn't expect a facter with weight == 0 to resolve over the builtin fact.

branan replied:
I'd say any fact with weight 0 should be considered "disabled". If that ends up surprising users in the future it's easy to fix :man_shrugging:

ekinanp commented:
jenkins please test this on windows10ent-64a,redhat7-64a

puppetcla commented:
CLA signed by all contributors.

ekinanp commented:
@McdonaldSeanp Thanks for double-checking that, will update the PR

branan commented on lib/src/ruby/fact.cc:
```diff
@@ -120,6 +120,21 @@ namespace facter { namespace ruby {
+
+                // None of our resolvers were able to resolve the value.
+                // Thus, we attempt to use the "built-in" fact's value.
+                auto builtin_value = facts[ruby.to_string(_name)];
```
Does it make more sense to do this at https://github.com/puppetlabs/facter/blob/b1960ad78d78616cf71a44164e2169801b15db27/lib/src/ruby/module.cc#L360 ?

branan commented on lib/src/ruby/fact.cc:
```diff
@@ -120,6 +120,21 @@ namespace facter { namespace ruby {
+                if (builtin_value) {
+                  // Already in collection, do not add
+                  add = false;
+                  _value = facter->to_ruby(builtin_value);
```
I don't really like wrapping the builtin fact as a ruby one - I'd much rather handle the fallback somewhere else, if we can

branan replied:
Ah, nevermind, I see we already do this wrapping for other cases when we don't add. Weird.

ekinanp commented:
jenkins please test this on windows10ent-64a,redhat7-64a

ekinanp commented:
It looks like we can't disable 0 weight facts because Facer will default custom facts that have no weight to have weight 0 (i.e. stuff breaks if 0 weight facts are disabled). I have, however, updated the PR so that built-in facts now have precedence over 0 weight facts, and have also updated the acceptance tests with the new behavior.

ekinanp commented on acceptance/tests/custom_facts/conflicts_with_builtin_fact.rb:
```diff
@@ -0,0 +1,106 @@
+
+    clear_custom_facts_on(agent, custom_fact_dir)
+
+    step "Verify that Facter uses the builtin fact's value when all conflicting custom facts fail to resolve" do
```
This test fails without the changes here, as expected.

ekinanp commented on acceptance/tests/custom_facts/conflicts_with_builtin_fact.rb:
```diff
@@ -0,0 +1,106 @@
+
+      clear_custom_facts_on(agent, custom_fact_dir)
+
+      step "when some custom facts have zero weight" do
```
This test also fails without the changes here, as expected. All other tests pass, as expected.

ekinanp commented:
@McdonaldSeanp Builtin facts already took precedence over custom facts that didn't specify a weight (because those facts defaulted to having a weight of 0) -- see https://github.com/puppetlabs/facter/blob/master/lib/src/ruby/fact.cc#L90. However, if one of those custom facts did have a weight, then builtin facts would not take precedence over the weightless custom facts. There should be some acceptance tests outlining these two cases. This change would break the latter scenario.
